### PR TITLE
implemented a parameter for skipping errors during model training in the backtest process

### LINKED ISSUE #9671: Avoid exception if a coin misses data while backtesting in FreqAI mode
## Describe your environment
(if applicable)

  * Operating system: ubuntu server
  * Python Version: 3.9.0
  * CCXT version: 4.1.74
  * Freqtrade Version: tag [2023.12](https://github.com/freqtrade/freqtrade/releases/tag/2023.12) from freqai docker image

## Describe the enhancement

Context: An hybrid freqtrade instance using freqAI decisions on enter/exit signals (to make sure my signals have the maximum potential)

Is it possible to NOT thow exception while a crypto is missing data on backtest mode but "just" do not take in account this coin. I use all cryptos from binance during backtest (to measure performance of several versions of a strategy using bullish, bearish and stable trend over a ful list of crypto to make decisions on production) .

If i have to check each separate coin data before manually, it will take me ages to process and i want to automate this thing properly over time regrding market changes.

I have a python script that automate backtest over several versions of my strategy to ensure  performance increase over each commit changes.
(i confess it takes days and days to process, but the result is quite amazing. FreqAI could be usefull to add a new brick about automated decision making)

For instance i have several exceptions occuring :

 - One from missing data
```
backtest_freqai_training            | 2024-01-11 20:46:44,388 - freqtrade.freqai.base_models.BaseClassifierModel - INFO - -------------------- Starting training LQTY/USDT --------------------
backtest_freqai_training            | 2024-01-11 20:46:44,391 - freqtrade.freqai.data_kitchen - INFO - LQTY/USDT: dropped 0 training points due to NaNs in populated dataset 0.
backtest_freqai_training            | 2024-01-11 20:46:44,391 - freqtrade.freqai.freqai_interface - WARNING - Training LQTY/USDT raised exception OperationalException. Message: LQTY/USDT: all training data dropped due to NaNs. You likely did not download enough training data prior to your backtest timerange. Hint:
backtest_freqai_training            | https://www.freqtrade.io/en/stable/freqai-running/#downloading-data-to-cover-the-full-backtest-period, skipping.
backtest_freqai_training            | Traceback (most recent call last):
backtest_freqai_training            |   File "/freqtrade/freqtrade/freqai/freqai_interface.py", line 351, in start_backtesting
backtest_freqai_training            |     self.model = self.train(dataframe_train, pair, dk)
backtest_freqai_training            |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/freqai/base_models/BaseClassifierModel.py", line 41, in train
backtest_freqai_training            |     features_filtered, labels_filtered = dk.filter_features(
backtest_freqai_training            |                                          ^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/freqai/data_kitchen.py", line 248, in filter_features
backtest_freqai_training            |     raise OperationalException(
backtest_freqai_training            | freqtrade.exceptions.OperationalException: LQTY/USDT: all training data dropped due to NaNs. You likely did not download enough training data prior to your backtest timerange. Hint:
backtest_freqai_training            | https://www.freqtrade.io/en/stable/freqai-running/#downloading-data-to-cover-the-full-backtest-period
backtest_freqai_training            | 2024-01-11 20:46:44,392 - freqtrade.freqai.freqai_interface - INFO - Saving metadata to disk.
backtest_freqai_training            | 2024-01-11 20:46:44,392 - freqtrade - ERROR - Fatal exception!
backtest_freqai_training            | Traceback (most recent call last):
backtest_freqai_training            |   File "/freqtrade/freqtrade/main.py", line 42, in main
backtest_freqai_training            |     return_code = args['func'](args)
backtest_freqai_training            |                   ^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/commands/optimize_commands.py", line 58, in start_backtesting
backtest_freqai_training            |     backtesting.start()
backtest_freqai_training            |   File "/freqtrade/freqtrade/optimize/backtesting.py", line 1392, in start
backtest_freqai_training            |     min_date, max_date = self.backtest_one_strategy(strat, data, timerange)
backtest_freqai_training            |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/optimize/backtesting.py", line 1309, in backtest_one_strategy
backtest_freqai_training            |     preprocessed = self.strategy.advise_all_indicators(data)
backtest_freqai_training            |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/strategy/interface.py", line 1346, in advise_all_indicators
backtest_freqai_training            |     return {pair: self.advise_indicators(pair_data.copy(), {'pair': pair}).copy()
backtest_freqai_training            |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/strategy/interface.py", line 1346, in <dictcomp>
backtest_freqai_training            |     return {pair: self.advise_indicators(pair_data.copy(), {'pair': pair}).copy()
backtest_freqai_training            |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/strategy/interface.py", line 1378, in advise_indicators
backtest_freqai_training            |     return self.populate_indicators(dataframe, metadata)
backtest_freqai_training            |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/user_data/strategies/abstract.py", line 289, in populate_indicators
backtest_freqai_training            |     df = super().populate_indicators(dataframe, metadata)
backtest_freqai_training            |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/user_data/strategies/freqai_predictor.py", line 168, in populate_indicators
backtest_freqai_training            |     dataframe = self.freqai.start(dataframe, metadata, self)
backtest_freqai_training            |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/freqai/freqai_interface.py", line 161, in start
backtest_freqai_training            |     dk = self.start_backtesting(dataframe, metadata, self.dk, strategy)
backtest_freqai_training            |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/freqai/freqai_interface.py", line 368, in start_backtesting
backtest_freqai_training            |     self.dd.save_metadata(dk)
backtest_freqai_training            |   File "/freqtrade/freqtrade/freqai/data_drawer.py", line 457, in save_metadata
backtest_freqai_training            |     dk.data["training_features_list"] = list(dk.data_dictionary["train_features"].columns)
backtest_freqai_training            |                                              ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
backtest_freqai_training            | KeyError: 'train_features'
```

 - And one from a past model exception i don't understand quite well:

```
backtest_freqai_training            | 2024-01-11 20:46:44,388 - freqtrade.freqai.base_models.BaseClassifierModel - INFO - -------------------- Starting training LQTY/USDT --------------------
backtest_freqai_training            | 2024-01-11 20:46:44,391 - freqtrade.freqai.data_kitchen - INFO - LQTY/USDT: dropped 0 training points due to NaNs in populated dataset 0.
backtest_freqai_training            | 2024-01-11 20:46:44,391 - freqtrade.freqai.freqai_interface - WARNING - Training LQTY/USDT raised exception OperationalException. Message: LQTY/USDT: all training data dropped due to NaNs. You likely did not download enough training data prior to your backtest timerange. Hint:
backtest_freqai_training            | https://www.freqtrade.io/en/stable/freqai-running/#downloading-data-to-cover-the-full-backtest-period, skipping.
backtest_freqai_training            | Traceback (most recent call last):
backtest_freqai_training            |   File "/freqtrade/freqtrade/freqai/freqai_interface.py", line 351, in start_backtesting
backtest_freqai_training            |     self.model = self.train(dataframe_train, pair, dk)
backtest_freqai_training            |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/freqai/base_models/BaseClassifierModel.py", line 41, in train
backtest_freqai_training            |     features_filtered, labels_filtered = dk.filter_features(
backtest_freqai_training            |                                          ^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/freqai/data_kitchen.py", line 248, in filter_features
backtest_freqai_training            |     raise OperationalException(
backtest_freqai_training            | freqtrade.exceptions.OperationalException: LQTY/USDT: all training data dropped due to NaNs. You likely did not download enough training data prior to your backtest timerange. Hint:
backtest_freqai_training            | https://www.freqtrade.io/en/stable/freqai-running/#downloading-data-to-cover-the-full-backtest-period
backtest_freqai_training            | 2024-01-11 20:46:44,392 - freqtrade.freqai.freqai_interface - INFO - Saving metadata to disk.
backtest_freqai_training            | 2024-01-11 20:46:44,392 - freqtrade - ERROR - Fatal exception!
backtest_freqai_training            | Traceback (most recent call last):
backtest_freqai_training            |   File "/freqtrade/freqtrade/main.py", line 42, in main
backtest_freqai_training            |     return_code = args['func'](args)
backtest_freqai_training            |                   ^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/commands/optimize_commands.py", line 58, in start_backtesting
backtest_freqai_training            |     backtesting.start()
backtest_freqai_training            |   File "/freqtrade/freqtrade/optimize/backtesting.py", line 1392, in start
backtest_freqai_training            |     min_date, max_date = self.backtest_one_strategy(strat, data, timerange)
backtest_freqai_training            |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/optimize/backtesting.py", line 1309, in backtest_one_strategy
backtest_freqai_training            |     preprocessed = self.strategy.advise_all_indicators(data)
backtest_freqai_training            |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/strategy/interface.py", line 1346, in advise_all_indicators
backtest_freqai_training            |     return {pair: self.advise_indicators(pair_data.copy(), {'pair': pair}).copy()
backtest_freqai_training            |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/strategy/interface.py", line 1346, in <dictcomp>
backtest_freqai_training            |     return {pair: self.advise_indicators(pair_data.copy(), {'pair': pair}).copy()
backtest_freqai_training            |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/strategy/interface.py", line 1378, in advise_indicators
backtest_freqai_training            |     return self.populate_indicators(dataframe, metadata)
backtest_freqai_training            |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/user_data/strategies/abstract.py", line 289, in populate_indicators
backtest_freqai_training            |     df = super().populate_indicators(dataframe, metadata)
backtest_freqai_training            |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/user_data/strategies/freqai_predictor.py", line 168, in populate_indicators
backtest_freqai_training            |     dataframe = self.freqai.start(dataframe, metadata, self)
backtest_freqai_training            |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/freqai/freqai_interface.py", line 161, in start
backtest_freqai_training            |     dk = self.start_backtesting(dataframe, metadata, self.dk, strategy)
backtest_freqai_training            |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backtest_freqai_training            |   File "/freqtrade/freqtrade/freqai/freqai_interface.py", line 368, in start_backtesting
backtest_freqai_training            |     self.dd.save_metadata(dk)
backtest_freqai_training            |   File "/freqtrade/freqtrade/freqai/data_drawer.py", line 457, in save_metadata
backtest_freqai_training            |     dk.data["training_features_list"] = list(dk.data_dictionary["train_features"].columns)
backtest_freqai_training            |                                              ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
backtest_freqai_training            | KeyError: 'train_features'
```

Could you enlight my candle about this please ?

Thanks a lot guys

### PR DESCRIPTION
**Pull request by xsa-dev**
## Summary

In case of any data disruptions related to a coin, currently, a hard exception occurs, interrupting the backtesting and limiting the completion of backtesting for the FreqAI strategy with other coins. To bypass this, it was suggested to use an exception skipping parameter as a warning, during which the exception is ignored during backtesting, allowing the backtesting to continue. In a separate case (with another parameter), resolving this issue may involve additional data preprocessing, which is not included in this feature's development.

Solve the issue: #9671

## Quick changelog

- new parameter `warn_exceptions_on_backtest_only` to `config_examples/config_freqai.example.json` in `config.freqai` 
- modify backtesting logic in `freqtrade/freqai/freqai_interface.py`

## What's new?

We added an exception skipping parameter to handle data disruptions during backtesting, letting the FreqAI strategy proceed across coins even when exceptions occur, and there's potential for further data preprocessing to tackle such issues later on.

### CODE REVIEW

File: freqtrade/freqai/freqai_interface.py [hunk 1]
```diff
@@ -354,25 +357,43 @@ def start_backtesting(
                         logger.warning(
                             f"Training {pair} raised exception {msg.__class__.__name__}. "
                             f"Message: {msg}, skipping.", exc_info=True)
+
+                        if self.new_feature_selector:
```
Comment (robcaulk) [inline freqtrade/freqai/freqai_interface.py]: What is the reason for putting this behind a flag? As far as I can tell, the proposed skip feature would be an improvement, so I don't see why anyone would opt out of it. Unless I am missing something?

File: freqtrade/freqai/freqai_interface.py [hunk 2]
```diff
@@ -354,25 +357,43 @@ def start_backtesting(
...
                     self.model = self.dd.load_data(pair, dk)
 
-                pred_df, do_preds = self.predict(dataframe_backtest, dk)
-                append_df = dk.get_predictions_to_append(pred_df, do_preds, dataframe_backtest)
-                dk.append_predictions(append_df)
-                dk.save_backtesting_prediction(append_df)
+                if hard_check_model_valid:
+                    pred_df, do_preds = self.predict(dataframe_backtest, dk)
+                    append_df = dk.get_predictions_to_append(pred_df, do_preds, dataframe_backtest)
+                    dk.append_predictions(append_df)
```
Comment (robcaulk) [inline freqtrade/freqai/freqai_interface.py:403]: The return dataframe (the DF the leaves `start_backtesting`) needs to match the length of the incoming dataframe length (the DF that enters `start_backtesting`). If I understand your goal here is to simply skip appending anything to the return dataframe. This would result in negative consequences back in the strategy. 

So to properly handle this, you would need some functionality to append some NaN or zeross in place of the desired predictions. We also need to ensure that this specific failed range has `do_predict=0` so that the strategy won't try to act on the 0s that we are filling here.

We need to keep in mind that the failure of training could be due to a lack of data *or* some other failure in the training. It would be good to distinguish these and log them so the user knows why things are failing but continuing. We can do a check for data length before even trying to train, if it is 0, then we can flag right there that there is no data available for training for this particular time-range. Otherwise, we can try the train and catch exception to report an issue that occurs in a particular training.

Please follow up with any questions or clarifications you have regarding this points, to avoid confusion.

**Review by robcaulk — CHANGES_REQUESTED** [review]

File: freqtrade/freqai/freqai_interface.py (hunk 1 above)
Comment (xsa-dev) [reply freqtrade/freqai/freqai_interface.py]: I did this as a flag because training on data with gaps is not standard behaviour, usually for training there should be all data. A more stable model is built on a stable dataset with no skips - then the usual failure (as in a train) will occur. For the more complex case, options are provided for the advanced user.

File: freqtrade/freqai/freqai_interface.py (hunk 2 above)
Comment (xsa-dev) [reply freqtrade/freqai/freqai_interface.py:403]: Do I understand correctly that you mean the following conversions:
- Adding NaN (Not a Number) in the place of expected predictions if the length of the returned dataframe is less than the length of the incoming dataframe.
- Adding zeros (0) to the place of expected predictions if the length of the returned dataframe is less than the length of the incoming dataframe.
- Setting the do_predict parameter to 0 for a failed range so that the strategy does not attempt to act on filled zeros.

The most important thing is to add NaN or zeros to the dataframe so that it has a length equal to the length of the incoming dataframe and zeros to distinguish between empty data and mark the training dataframe set to 0 if there are no other predictions? I could make multiple options for processing the dataset, not just true or false.

File: freqtrade/freqai/freqai_interface.py (hunk 1 above)
Comment (robcaulk) [reply freqtrade/freqai/freqai_interface.py]: It appears that there is some confusion of how `start_backtesting()` works. Let me try to explain it in plain english:

It:
0) gets a full `--timerange` of data for a single coin from the strategy.
1) splits the `--timerange` up according to `--timerange` / `backtest_period_days` (the user desired backtest timerange divided by the `backtest_period_days` parameter.) Let's say this ends up being 10 sub-timeranges.
2) for each sub timerange, it trains a model.
3) It uses that model make an inference on the subsequent `backtest_period_days`. 
4) It appends these predictions to the return dataframe. 
5) move to next timerange and repeat steps 2-4. 
6) do this until there are no more timeranges.

This means that it is training X separate models, where X is `--timerange`/`backtest_period_days`

The issue that your PR is aiming to alleviate is to allow one of those model trainings to fail. That may be due to lack of data or may be due to some other problem in the training procedure. 

So that means that there will be no model trained with "gaps" in data as you are suggesting. The only thing that occurs is if one of the sub timerange model trainings fails, then that entire sub timerange `backtest_period_days` will return some filler + `do_predict = 0` so that the user strategy doesnt act on anything for that sub timerange.

This has nothing to do with a model getting trained on data gaps. Therefore, there is no reason to be hiding this functionality behind a flag.

File: freqtrade/freqai/freqai_interface.py (hunk 2 above)
Comment (robcaulk) [reply freqtrade/freqai/freqai_interface.py:403]: Have a look at my explanation of the `start_backtesting()` procedure. 

This is not about checking if the length of the returned dataframe is less than the length of the incoming dataframe. 

This is about ensuring that for each iteration of a sub-timerange, you always append the same amount of candles, whether those candles are real predictions from a properly trained model, or filler predictions from a failure. Same number appended, then moving to next sub-timerange.

Comment (xsa-dev) [reply freqtrade/freqai/freqai_interface.py:403]: ### What Needs to Be Done:
- Match the lengths of dataframes: Ensure that the length of the return dataframe matches the length of the incoming dataframe. This can be achieved by adding NaN or zeros in place of expected predictions if the number of predictions is less than required.
- Handle failures: If there is a failure during training, such as due to a lack of data or other issues, it should be handled appropriately.
- Set do_predict=0 for failed ranges so that the strategy does not attempt to act on filled zeros.
- Log failures so that users are aware of the reason for the training failure.
- Distinguish between types of failures: Differentiate between different types of failures (e.g., lack of data versus other issues) and log them accordingly.
- Check data length before training: Before attempting to train the model, check the length of the incoming data. If there is no data, immediately indicate the issue and avoid training attempts.
- Consistency and robustness: Ensure consistency in the process across different iterations and handle data anomalies appropriately.
### What Should Not Be Done:
- Do not ignore errors: Ignoring errors during the backtesting process can lead to inaccurate results and incorrect decision-making.
- Do not skip error logging: Being unaware of what happens during training or testing can make diagnosing and fixing problems in the future difficult.
- Do not alter the length of the incoming dataframe: The length of the incoming dataframe should not be changed; all manipulations should be done only with the return dataframe.
- Do not overdo detailing: Although detailing is important for understanding the process, excessive detailing can complicate implementation.

Did I understand correctly?

**Comment by xsa-dev** [comment]
Suggested edit:
```diff
diff --git a/freqtrade/freqai/freqai_interface.py b/freqtrade/freqai/freqai_interface.py
index 224b9dbe1..ae17c5940 100644
--- a/freqtrade/freqai/freqai_interface.py
+++ b/freqtrade/freqai/freqai_interface.py
@@ -5,7 +5,7 @@ from abc import ABC, abstractmethod
 from collections import deque
 from datetime import datetime, timezone
 from pathlib import Path
-from typing import Any, Dict, List, Literal, Optional, Tuple, Union
+from typing import Any, Dict, List, Literal, Optional, Tuple
 
 import datasieve.transforms as ds
 import numpy as np
@@ -108,7 +108,7 @@ class IFreqaiModel(ABC):
         self.data_provider: Optional[DataProvider] = None
         self.max_system_threads = max(int(psutil.cpu_count() * 2 - 2), 1)
         self.can_short = True  # overridden in start() with strategy.can_short
-        self.model: Union[Any, None] = None
+        self.model: Any = None
         if self.ft_params.get('principal_component_analysis', False) and self.continual_learning:
             self.ft_params.update({'principal_component_analysis': False})
             logger.warning('User tried to use PCA with continual learning. Deactivating PCA.')
@@ -258,7 +258,7 @@ class IFreqaiModel(ABC):
                     self.dd.save_metric_tracker_to_disk()
 
 
-    def _train_model(self, dataframe_train, pair, dk, tr_backtest) -> Union[Any, None]:
+    def _train_model(self, dataframe_train, pair, dk, tr_backtest):
         try:
             self.tb_logger = get_tb_logger(self.dd.model_type, dk.data_path,
                                             self.activate_tensorboard)

```

**Review by xsa-dev — COMMENTED** [review]
Done

File: freqtrade/freqai/freqai_interface.py [hunk 3]
```diff
@@ -108,7 +108,6 @@ def __init__(self, config: Config) -> None:
         self.data_provider: Optional[DataProvider] = None
         self.max_system_threads = max(int(psutil.cpu_count() * 2 - 2), 1)
         self.can_short = True  # overridden in start() with strategy.can_short
-        self.model: Any = None
```
Comment (robcaulk) [inline freqtrade/freqai/freqai_interface.py]: What is the reason for removing this attribute and default setting?

**Comment by xsa-dev** [comment]
close https://github.com/freqtrade/freqtrade/issues/10145

**Comment by xmatthias** [comment]
@robcaulk is there anything missing in this PR? or is it something we ain't going to accept so it can be closed?

**Review by robcaulk — CHANGES_REQUESTED** [review]

File: freqtrade/freqai/freqai_interface.py (hunk 3 above)
Comment (xsa-dev) [reply freqtrade/freqai/freqai_interface.py]: You are looking at an outdated file; my branch definitely has this attribute: `self.model: Any = None.`
Or am I mistaken?

**Comment by xmatthias** [comment]
I've decided to go with #13143 instead - as it has fewer side-effects and doesn't require explicit opt in.

I'll hence close this - as the underlying bug has been fixed.
